### PR TITLE
add benchmark 'serialization crate v0.2.8'

### PR DESCRIPTION
**Pull request by Bruce0203**

### CODE REVIEW

**Comment by caibear** [comment]
From reading serialization's code I see the effort to make it fast.

That being said, to accurately represent it on these benchmarks, I think these unchecked operations need to be checked or the deserialize benchmark needs to be labeled as unvalidated (like some of rkyv's benchmarks).

[not checked if valid utf8](https://github.com/Bruce0203/serialization/blob/63315a2f190bccb646e30456dc3597a5bd7e59ca/src/impls.rs#L423)
[not checked if valid utf8](https://github.com/Bruce0203/serialization/blob/63315a2f190bccb646e30456dc3597a5bd7e59ca/serialization_minecraft/src/lib.rs#L240)
[not checked if not 0 or 1](https://github.com/Bruce0203/serialization/blob/63315a2f190bccb646e30456dc3597a5bd7e59ca/serialization_minecraft/src/lib.rs#L231)
[not checked if 0](https://github.com/Bruce0203/serialization/blob/63315a2f190bccb646e30456dc3597a5bd7e59ca/src/impls.rs#L461)
[not checked if 0](https://github.com/Bruce0203/serialization/blob/63315a2f190bccb646e30456dc3597a5bd7e59ca/src/impls.rs#L406)

These are all the unchecked deserialize impls I found by briefly reading the code, but I may have missed some.

**Comment by mumbleskates** [comment]
in addition to that, for pure sanity checks we also want to make sure that the bench function contains an `assert_eq!` for at least one round-trip. when i added this it caught at least one case where a serializer benchmark wasn't actually encoding a large chunk of the data, so it's good to have for any encoder/decoder that has a convenient path back to input `T`.

**Comment by Bruce0203** [comment]
Thank you for interest

[not checked if valid utf8](https://github.com/Bruce0203/serialization/blob/63315a2f190bccb646e30456dc3597a5bd7e59ca/src/impls.rs#L423) -> [fixed](https://github.com/Bruce0203/serialization/blob/c84ba0b5e8005df02b08d9717625e43a2d76bdfa/src/impls.rs#L410-L412)
[not checked if valid utf8](https://github.com/Bruce0203/serialization/blob/63315a2f190bccb646e30456dc3597a5bd7e59ca/serialization_minecraft/src/lib.rs#L240) -> [fixed](https://github.com/Bruce0203/serialization/blob/c84ba0b5e8005df02b08d9717625e43a2d76bdfa/serialization_minecraft/src/lib.rs#L261-L263)
[not checked if not 0 or 1](https://github.com/Bruce0203/serialization/blob/63315a2f190bccb646e30456dc3597a5bd7e59ca/serialization_minecraft/src/lib.rs#L231) -> [fixed](https://github.com/Bruce0203/serialization/blob/c84ba0b5e8005df02b08d9717625e43a2d76bdfa/serialization_minecraft/src/lib.rs#L248-L250)
[not checked if 0](https://github.com/Bruce0203/serialization/blob/63315a2f190bccb646e30456dc3597a5bd7e59ca/src/impls.rs#L461) -> [fixed](https://github.com/Bruce0203/serialization/blob/c84ba0b5e8005df02b08d9717625e43a2d76bdfa/src/impls.rs#L453-L455)
[not checked if 0](https://github.com/Bruce0203/serialization/blob/63315a2f190bccb646e30456dc3597a5bd7e59ca/src/impls.rs#L406) -> [fixed](https://github.com/Bruce0203/serialization/blob/c84ba0b5e8005df02b08d9717625e43a2d76bdfa/src/impls.rs#L388-L390)


[assertion after benchmark](https://github.com/Bruce0203/rust_serialization_benchmark/blob/master/src/bench_serialization.rs#L35)

**Comment by mumbleskates** [comment]
some cursory thoughts about some of the safety/validation there, particularly the utf8:

the standard library's `core::str::from_utf8` validation function is already [quite fast and sophisticated](https://doc.rust-lang.org/src/core/str/validations.rs.html#209-233), and will not fall back to returning false errors for any non-ascii-but-valid-utf8 text. it looks like your replacement, `is_ascii_simd`, *does* do that: it only seems to check for the presence of any MSB byte by byte, in a way that invites SIMD optimization (though the simd chunk size is very large at 128 bytes, and it seems you do not check that the chunks are aligned in any way, which may in the end rule them out for simd).

Or rather, it's not actually *trying* to validate utf-8, nor does it pretend to. it's checking if the passed bytes are ascii: it does exactly what it says on the tin; arguably, the problem with false errors really happens later, when we assume that [anything that isn't ascii is invalid data when validating decoded Strings](https://github.com/Bruce0203/serialization/blob/c84ba0b5e8005df02b08d9717625e43a2d76bdfa/src/impls.rs#L411).

have you looked at using the standard library validation already? have you compared its performance to `is_ascii_simd`? i know the code is a lot heavier there so i'm kind of curious what the difference is. Personally i'd opt for real validation because failing to round-trip anything past ascii would end up feeling very stuck in the past, but that's probably just my usecases talking. If there is a real performance difference here i agree with @caibear that it should probably go under a "reduced validation / unvalidated" umbrella for performance caveats.

**Comment by mumbleskates** [comment]
also to note: i don't think the `benchmark` gh-action is usually run in branches that add new benchmarks like this; usually running `cargo test --benches` in the branch and the `build` gh-action tests should be enough to show that benches are working.

**Comment by caibear** [comment]
I don't think `is_ascii_simd` will improve performance for the serialization crate on this benchmark since all the strings are < 128 bytes. Also std's `<[u8]>::is_ascii` and `str::from_utf8` should also perform about the same since they both check usizes in a loop without simd.

The `is_ascii_simd` function appears to be from [bitcode](https://github.com/SoftbearStudios/bitcode/blob/83dc478963bcff864e6d8385d2987fa9f93a9e29/src/str.rs#L142-L160) where I used it to check the validity of thousands of short strings at once (bitcode concatenates them to form 1 large string). In this case, it is a clear performance win since on large strings 
`is_ascii_simd` bottlenecks on ram speed while `<[u8]>::is_ascii` and `str::from_utf8` don't. I should also point out that bitcode has a [fallback](https://github.com/SoftbearStudios/bitcode/blob/83dc478963bcff864e6d8385d2987fa9f93a9e29/src/str.rs#L92-L112) for non-ascii strings that's still faster than `str::from_utf8` on every short string.

For long string validation (not measured on this benchmark) there's the [simdutf8](https://crates.io/crates/simdutf8) crate which is just as fast as `is_ascii_simd`. I didn't use it for bitcode because bitcode's fastest path only works on ascii.

TLDR: serialization should use `str::from_utf8` since it won't benefit from this bitcode-specific optimization. If increased performance on long strings is desired there is the [simdutf8](https://crates.io/crates/simdutf8) crate.

**Comment by caibear** [comment]
> also to note: i don't think the `benchmark` gh-action is usually run in branches that add new benchmarks like this; usually running `cargo test --benches` in the branch and the `build` gh-action tests should be enough to show that benches are working.

If you do want to run the benchmarks on your branch, you can manually run the `bench` workflow in `your_fork -> Actions -> bench -> run workflow -> your_branch`. I'd be curious to see the results of the extensive optimizations you've made in serialization.

**Comment by caibear** [comment]
> Thank you for interest
> 
> [not checked if valid utf8](https://github.com/Bruce0203/serialization/blob/63315a2f190bccb646e30456dc3597a5bd7e59ca/src/impls.rs#L423) -> [fixed](https://github.com/Bruce0203/serialization/blob/c84ba0b5e8005df02b08d9717625e43a2d76bdfa/src/impls.rs#L410-L412) [not checked if valid utf8](https://github.com/Bruce0203/serialization/blob/63315a2f190bccb646e30456dc3597a5bd7e59ca/serialization_minecraft/src/lib.rs#L240) -> [fixed](https://github.com/Bruce0203/serialization/blob/c84ba0b5e8005df02b08d9717625e43a2d76bdfa/serialization_minecraft/src/lib.rs#L261-L263) [not checked if not 0 or 1](https://github.com/Bruce0203/serialization/blob/63315a2f190bccb646e30456dc3597a5bd7e59ca/serialization_minecraft/src/lib.rs#L231) -> [fixed](https://github.com/Bruce0203/serialization/blob/c84ba0b5e8005df02b08d9717625e43a2d76bdfa/serialization_minecraft/src/lib.rs#L248-L250) [not checked if 0](https://github.com/Bruce0203/serialization/blob/63315a2f190bccb646e30456dc3597a5bd7e59ca/src/impls.rs#L461) -> [fixed](https://github.com/Bruce0203/serialization/blob/c84ba0b5e8005df02b08d9717625e43a2d76bdfa/src/impls.rs#L453-L455) [not checked if 0](https://github.com/Bruce0203/serialization/blob/63315a2f190bccb646e30456dc3597a5bd7e59ca/src/impls.rs#L406) -> [fixed](https://github.com/Bruce0203/serialization/blob/c84ba0b5e8005df02b08d9717625e43a2d76bdfa/src/impls.rs#L388-L390)
> 
> [assertion after benchmark](https://github.com/Bruce0203/rust_serialization_benchmark/blob/master/src/bench_serialization.rs#L35)

I noticed that you transmute between `NonMaxU32` and `u32`. Normally this implementation would work however `NonMaxU32` is [backed by a `NonZeroU32`](https://github.com/LPGhatguy/nonmax/blob/645abf30102dd880542865195c36893ee89994b5/src/lib.rs#L126) so the niche is all zeros, not all ones. When the u32 value is retrieved by `NonZeroU32::get` it's [xored](https://github.com/LPGhatguy/nonmax/blob/645abf30102dd880542865195c36893ee89994b5/src/lib.rs#L154) with all ones to fix this.

Rather than changing [your code](https://github.com/Bruce0203/serialization/blob/a023e0d268c86f3cea5e06680897ffb0a44a399a/src/impls.rs#L452) to check if the value is 0, you could use the [NonMaxU32::new](https://github.com/LPGhatguy/nonmax/blob/645abf30102dd880542865195c36893ee89994b5/src/lib.rs#L132) function to avoid relying on implementation details.

**Comment by caibear** [comment]
One last comment for now. I noticed some [unsound optimizations](https://github.com/Bruce0203/serialization/blob/a023e0d268c86f3cea5e06680897ffb0a44a399a/src/impls.rs#L74-L112) that allow `Option`/`Result` to be decoded in place. This kind of code heavily relies on the current compiler internals and also doesn't work for types with niches like `Option<&T>`.

I'm somewhat sympathetic to unsound optimizations in the pursuit of being faster than other libraries having used [some](https://github.com/SoftbearStudios/bitcode/blob/83dc478963bcff864e6d8385d2987fa9f93a9e29/src/derive/vec.rs#L45-L79) myself in bitcode. I even considered this exact optimization in bitcode, but I felt it was a little too far. I think it would be best to work on getting this functionality implemented into the Rust std library (e.g. function/macro to write discriminant of uninit enum).

**Comment by mumbleskates** [comment]
it might be slightly off topic but i wonder @caibear whether the head-of-slice alignment for `is_ascii_simd` ever mattered: whether the compiler is happy to use unaligned-reader simd intrinsics instead(?), whether it requires alignment and it just happens to line up at runtime, or whether there's something about the chunks call that i missed that makes this always work

**Comment by caibear** [comment]
@mumbleskates I only benchmarked it on x86-64 and the compiler is happy to use unaligned simd load instructions. There is probably a slight performance cost to them but not enough to overshadow the gain over scalar code. I didn't bother making it more complex since the algorithm already runs faster than ram so it can't be improved on large datasets which don't fit into the cpu cache.
